### ISSUE TITLE
Not able to replicate XSUM results

### ISSUE THREAD

**Issue by jpilaul** [XiangLi1999/PrefixTuning#2]
Thanks for making your code public. I am having difficulties replicating XSUM results. I think that it could be a data processing issue. Can you provide more details on XSUM data and preprocessing.

I got the data from here:
http://bollin.inf.ed.ac.uk/public/direct/XSUM-EMNLP18-Summary-Data-Original.tar.gz

Then, I tried to preprocess it myself since you use a `.len` file in https://github.com/XiangLi1999/PrefixTuning/blob/cleaned/seq2seq/utils.py#L104-L106
``` python
self.src_file = Path(data_dir).joinpath(type_path + ".source")
self.tgt_file = Path(data_dir).joinpath(type_path + ".target")
self.len_file = Path(data_dir).joinpath(type_path + ".len")
```

Still, the results are significantly lower than in the paper.
Thanks

**Comment by XiangLi1999** [comment]
I used the huggingface data (now in the legacy file). 

wget https://cdn-datasets.huggingface.co/summarization/xsum.tar.gz

tar -xzvf xsum.tar.gz

**Comment by jpilaul** [comment]
Thanks. I still need a `.len` file. Was that the word count of the source or target?

**Comment by XiangLi1999** [comment]
I think you can run the code without the .len file, the method builds on top of the  abstract class (I think).

**Comment by jpilaul** [comment]
Hi again, I reran the model and I am still getting the following eval scores (from `metrics.json`):

```sql
     "val_avg_loss": 7.211210250854492,
     "val_avg_rouge1": 16.93965,
     "val_avg_rouge2": 3.05505,
     "val_avg_rougeL": 13.90935,
     "val_avg_gen_time": 1.0583273768424988
```

Which is very far from the reported scores...

I am using exactly the same commands as in the `README`.

**Comment by XiangLi1999** [comment]
It's quite clear that your training collapses in this setting, val loss should be much lower. Try a smaller learning rate or a larger effective batchsize. 

You could probably  try this first! 
`python train_bart.py --mode xsum --preseqlen 40 --do_train yes --fp16 yes --bsz 16  --epoch 15  --gradient_accumulation_step 3 --learning_rate 0.00005  --mid_dim 800`

**Comment by jpilaul** [comment]
So the hyper parameters in the `README` are the not the ones that you used to get the results in the paper?

**Comment by XiangLi1999** [comment]
the parameters in the README is the one I used, I am suggesting fewer epochs and smaller prefix length in the previous reply so that it's easier/quicker to train and debug any basic problem. 

Alternatively, https://worksheets.codalab.org/bundles/0xa3f0cd3c10c7490ab508a351968cbdcf could be a good reference to check against!

**Comment by jpilaul** [comment]
Thank you for the extra info. I have fixed the issue and now my `val_avg_loss` goes down nicely. It's actually at `1.51` at the end of epoch 30. You are at `1.59` according to your link. However, my rouge scores are still lower than yours by almost `2` percentage points. my `val_avg_gen_len` is also around 30 vs 35 for you. I had to lower my batch size which could explain the difference. I will add updates as I continue investigating the issue.

**Comment by jpilaul** [comment]
I found a hyperparameter that can explain the difference in the evaluation scores: `n_val` is set 500. However, this does not explain `2` percentage points difference in `R1` on the test set...

**Comment by jpilaul** [comment]
I tried with the same batch size and with `n_val` is set 500 and I am still `2` percentage points lower than your reported R1 and around `1` percentage point lower than your reported R2/RL.

**Comment by XiangLi1999** [comment]
Hi Jonathan, 

For validation ROUGE2 and ROUGE1, did you get something close to the CodaLab results of validation? (that's done with length normalization of 1. so you should probably get similar rouge2 on validation set.)

For test, another trick that I did and described in the paper is length normalization of 0.8. It's another hyper-parameter that you could pass to the decoder. 

Also, I used a slightly different evaluation scripts, the evaluation scripts takes calls the same rouge library, but takes the average of scores rather than bootstrap.

**Comment by jpilaul** [comment]
Hi Lisa,
I cannot replicate your results on validation using your code base and your reference's hyperparameters. I am getting the same training loss as you but a lower validation loss and Rouge scores still `2` percentage points lower than your reference. See here:
https://worksheets.codalab.org/bundles/0xa8f9ebe388914f4597f8cb3679f93c2c

**Comment by jpilaul** [comment]
I am closing the issue since I was able to replicate results with the altered xsum dataset (https://github.com/XiangLi1999/PrefixTuning/issues/20). On the original xsum dataset, I get 2-3 rouge points lower.

**Comment by zhaone** [comment]
@jpilaul @XiangLi1999  Hi, I got the same lower metrics (3 pt lower than that reported in the paper) on the original dataset. Where can I find the altered dataset? I followed the #20 but did not find the altered one. Should I add 3 `.len` files for train, val and test sets manually?

BTW: due to the GPU and CUDA version, I use a newer version PyTorch-lightning. Don't know whether this factor influence final metrics heavily?

**Comment by XiangLi1999** [comment]
the "xsum_news" split is intended for testing OOD (out-of-distribution) performance. It's here https://worksheets.codalab.org/bundles/0x58f85171b43f4e61bf411c35faab369d, the dir named xsum_news. 
The table2 in the paper still uses the original xsum split (the directory named xsum).

**Comment by jpilaul** [comment]
Reopening issue since I thought I was using the right dataset...

**Comment by jpilaul** [comment]
Using the standard BART hyperparameters ([fairseq link](https://github.com/pytorch/fairseq/blob/main/examples/bart/README.summarization.md#4-fine-tuning-on-cnn-dm-summarization-task)) for training `xsum_original`, I find that BART fully `finetuned` performs better than the scores shown in table 3  (added in comment below) on `xsum_news`. Here are my "finetune" results:
```python
FINE_TUNE = {
       "test_rouge1": 39.53074645996094,   # 39.20
       "test_rouge2": 16.876326751708984,  # 16.35
       "test_rougeL": 31.549299697875977,  # 31.15
}
```
As for  `prefix tuning` on `xsum_news`, my own run is still around 1 percentage points lower using your hyperparameters:
```python
PREFIX = {
       "test_rouge1": 38.87569885253906,  # 39.41
       "test_rouge2": 16.090474891662598, # 16.87
       "test_rougeL": 30.79502716064453,  # 31.47
}
```

**Edit**: I just realized that might have doubled the batch size which could explain the differences.
